#44 - GetCandy (#46)
blumilksoftware/internal-meetups: 2022-07-06-get-candy/slides/slides.md

> <h2>GetCandy - laravel <br> e-commerce package</h2>

Reviewer: ```suggestion
<h2>GetCandy - Laravel <br> e-commerce package</h2>
```

<img style="height: 8em" data-src="presentations/2022-07-06-get-candy/images/getcandy_icon.svg">

<p>Adrian Gawle 06.07.2022</p>

---

## E-commerce

<p>E-commerce (electronic commerce) is the buying and selling of goods and services,

> or transmitting of funds or data, over an electronic network, primarly the internet. These transactions occur either as:</p>

Reviewer: ```suggestion
or transmitting of funds or data, over an electronic network, primarily the internet. These transactions occur either as:</p>
```

<ul>
<li>B2B - business-to-business</li>
<li>B2C - business-to-consumer</li>
</ul>

---

## GetCandy

<p>

> An Opensource headless e-commerce package built on laravel. It provides a quite nice admin hub built on livewire.

Reviewer: ```suggestion
An opensource headless e-commerce package built on Laravel. It provides a quite nice admin hub built on Laravel Livewire.
```

First version of GetCandy was developed in 2018. In 2021, after their newly learnt experience, they decided to create version 2.
</p>

---

### Some numbers

- https://github.com/getcandy/getcandy
- Stars - 467
- Forks - 77
- Last commit - 6 days ago
- Last release - 2.0-beta13.2 - 2022-06-23

---

### Installation

#### Requirements:
<ul>
    <li>PHP ^8.0</li>
    <li>Laravel 8|9</li>
    <li>MySQL 5.7+ / PostgesSQL 9.2+</li>
    <li>PHP extensions: exif, intl and GD</li>
</ul>

---

```shell
composer require getcandy/admin
php artisan vendor:publish --tag=getcandy
```

---

> <p>As the GetCandy heavly uses default laravel user. We need to add GetCandyUser trait to our model.</p>

Reviewer: ```suggestion
<p>As the GetCandy heavily uses default Laravel user. We need to add GetCandyUser trait to our model.</p>
```

```php
use GetCandy\Base\Traits\GetCandyUser;
// ...

class User extends Authenticatable
{
    use GetCandyUser;
    // ...
}
```

---

### Search Configuration

<p>GetCandy uses Laravel Scout, for local development collection driver should be enough, but meilisearch driver is recommended.</p>

---

> <p>Example of meilisearch setup in docker</p>

Reviewer: ```suggestion
<p>Example of meilisearch setup in Docker</p>
```

        environment:
            - MEILI_MASTER_KEY=${MEILISEARCH_KEY}
        networks:
            - demo-store
        restart: unless-stopped
```

---

<p>Once we have meilisearch up and running we need to add two packages to our project</p>

```shell
composer require meilisearch/meilisearch-php http-interop/http-factory-guzzle
```

<p>and update our `.env` file:</p>

```
SCOUT_DRIVER=meilisearch
MEILISEARCH_HOST=demo-store-meilisearch:7700
MEILISEARCH_KEY=masterKey
```

---

### Admin Hub

<p>Publish Assets</p>

```shell
php artisan getcandy:hub:install
```

---

### Run Migrations
<p>GetCandy uses table prefix, so it won't interfere with existing tables. It also assumes that our User ID field is a "BIGINT". Both of these settings can be changed in config/getcandy/database.php</p>

---

<p>Then we just run migrate command</p>

```shell
php artisan migrate
```

---

<p>Next we run installer</p>

```shell
php artisan getcandy:install
```

<p>This command will do a few things:</p>

- Copy the config files if we want (usefull for restoring default config)
- Create a default admin user (if required)
- Seeding initial data (like Countries/Currencies/Default Tax etc.)

---

> <p>If we setted up meilisearch we need to run last command</p>

Reviewer: ```suggestion
<p>If we set up meilisearch we need to run last command</p>
```

```shell
php artisan getcandy:meilisearch:setup
```

> <p>It will generate any required indexes, filterable and sortable fields, that are needed for scout</p>

Reviewer: ```suggestion
<p>It will generate any required indexes, filterable and sortable fields, that are needed for Laravel Scout</p>
```

---

### Key features

- Custom Attributes
- Tags
- Autogenerated thumbnails / Image resizer
- Inventory managment
- Shipping
- Associations - cross-sells, up-sells, alternatives
- Multilanguage custom URLs for Products and Collection
- Multi-channels

---

<p>Let's take a look at the hub</p>

---

### Extendability

<p>Even at current state GetCandy is quite extendable</p>

---

#### Admin Hub

- Currently only side menu and settings menu for new links
- Slots - simple Livewire compontents that can be added to different pages. Must implement `AbstractSlot` interface.
- Tables - currently only `GetCandy\Facades\OrdersTable`

---

#### Cart Modifiers

> <p>If we need to modify a Cart before/after calculation took place. For this GetCandy uses Pipelines.</p>

Reviewer: ```suggestion
<p>If we need to modify a Cart this takes place before or after calculation. For this GetCandy uses Pipelines.</p>
```
